File: 110-balanced-binary-tree/balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def checkHeightAndBalance(node):
            if not node:
                return 0  # Height of an empty tree is 0

            # Recursively get the height of left and right subtrees
            leftHeight = checkHeightAndBalance(node.left)
            if leftHeight == -1:  # Left subtree is unbalanced
                return -1

            rightHeight = checkHeightAndBalance(node.right)
            if rightHeight == -1:  # Right subtree is unbalanced
                return -1

            # Check if the current node is balanced
            if abs(leftHeight - rightHeight) > 1:
                return -1  # Current node is unbalanced

            # Return the height of the current node
            return 1 + max(leftHeight, rightHeight)

        # If the function returns -1, the tree is unbalanced
        return checkHeightAndBalance(root) != -1

File: 110-balanced-binary-tree/balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    public int dfs(TreeNode root) {
        
        if (root == null) {
            return 0;
        }

        int left = dfs(root.left);
        int right = dfs(root.right);

        return 1 + Math.max(left, right);
    }

    public boolean isBalanced(TreeNode root) {

        if (root == null) {
            return true;
        }

        int left = dfs(root.left);
        int right = dfs(root.right);

        if (Math.abs(left - right) > 1) {
            return false;
        }

        boolean isBalancedLeft = isBalanced(root.left);
        boolean isBalancedRight = isBalanced(root.right);
        return isBalancedLeft && isBalancedRight;
    }
}